### PR TITLE
Make autocomplete work with entity ids

### PR DESCRIPTION
**Pull request by Kicu**
### Details
 - this PR implements autocomplete for every filter that is not "plain text" but requires saving some ids of the entity the user actually chosen (`from`, `to`, `in`, `tax`, `cardID`)
 - we save the user autocomplete choice and later substitute query with the entity id before sending request to backend (details below)
 - contains very small change to autocomplete parser, and 2 functions which I put into separate files to write tests for them
 - **important:** this is only expected to work in `SearchRouter`, `SearchPageHeader` will still work using old code; I will make this autocomplete work also on Results page in `SearchPageHeader` but in a separate PR. Doing it in here would make this PR grow way over 1k loc

-----

#### Detailed autocomplete description
 - Previously we were storing the full query that user typed verbatim
 - For example user would type `test from:john@expensify.com`, then when submitting we would run a function [standardizeQueryJSON](https://github.com/Expensify/App/blob/main/src/libs/SearchQueryUtils.ts#L607) to traverse the parsed tree and substitue all **plain text** values with id values for specific filters. (this helper fn did the job: https://github.com/Expensify/App/blob/main/src/libs/SearchQueryUtils.ts#L169)
- However this breaks for any case where the **plain text** value is non-unique. That is the case for chat/report names. For example when I type `in:adm` into my autocomplete I get more than 5 different rooms all named `#admins`. Previous solution couldn't handle a string like this `in:#admins hello` because we don't know which room `#admins` room it might be. This is the source of bug https://github.com/Expensify/App/issues/50976.

#### new implementation
 - the main source of truth is still the "plain text" text input value that the users sees
 - in addition any time a user picks (clicks) a choice from autocomplete list, we save data in a `substitutionsMap` object
 - when we need to process the query we can always correctly substitute a `string` that user sees in his query with an `autocompleteID` of the entity that can be stored in URL and sent to backend; these ids can be `accountID` for a user or `reportID` for chat/room/report, and some other ids for tags and cards
 - old `standardizeQueryJSON` function can soon be removed, but this mechanism described above now only works for `SearchRouter`, whereas input inside `SearchHeaderPage` still uses the old substitution mechanism, and thus needs the function
 - tests for `getQueryWithSubstitutions` and `getUpdatedSubstitutionsMap` should help understand this

### Fixed Issues

$ https://github.com/Expensify/App/issues/50976
$ https://github.com/Expensify/App/issues/50947
$ https://github.com/Expensify/App/issues/50944
$ https://github.com/Expensify/App/issues/50943
PROPOSAL:

### Tests
- open SearchRouter via clicking on 🔎 icon or `cmd +k` shortcut
- verify that when using contextual search (inside a room) room name is correctly displayed
- test autocomplete for `from:...`, `to:...`, `cardID:...` and `tax:...` 
- in every case user should see plain text display name, but backend should receive id instead

### QA Steps

- open SearchRouter via clicking on 🔎 icon or `cmd +k` shortcut
- verify that when using contextual search (inside a room) room name is correctly displayed
- test autocomplete for `from:...`, `to:...`, `cardID:...` and `tax:...` 
- in every case user should see plain text display name and the search should work

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/3f6de06e-4a05-44c8-a8ce-dddb69148b11

screen-autocomplete-andr

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/10bc31f0-f1d4-47b4-8780-3b2276f27880

</details>

<details>
<summary>iOS: mWeb Safari</summary>
ios

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/433f7d8c-bdc2-4115-87b4-2e1b2fd4a516

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Review by SzymczakJ — APPROVED** [review]
LGTM

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (luacmartins) [inline src/components/Search/SearchRouter/SearchRouter.tsx:250]: I think this is ok for now, but we might have to add some more details to the name to help users differentiate multiple reports with the same name, e.g. `admin` rooms

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (luacmartins) [inline src/components/Search/SearchRouter/SearchRouter.tsx:349]: NAB this variable seems unnecessary, we can just pass the param inline
```suggestion
            const computeNodeValueFn = SearchQueryUtils.getUpdatedAmountValue;
```

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (luacmartins) [inline src/components/Search/SearchRouter/SearchRouter.tsx:359]:
```suggestion
    const onAutocompleteSuggestionClick = (autocompleteKey: string, autocompleteID: string) => {
```

File: src/components/Search/SearchRouter/getQueryWithSubstitutions.ts
Comment (luacmartins) [inline src/components/Search/SearchRouter/getQueryWithSubstitutions.ts:4]: Do we need the value key here? AFAIK we always have the map as `{[filterKey:typedValue]: {value: [id]}}`. Could we simplify it to `{[filterKey:typedValue]: [id]}`?
```suggestion
type SubstitutionEntry = {value: string};
```

**Review by luacmartins — COMMENTED** [review]
Looking great! Left a few small comments. 

NAB: @Expensify/design we might wanna find a way to differentiate between some of these reports. When I search for `in:#admin` I get a list of different reports, but I have no way of knowing which one is which. Maybe we can append the policy name to it?

Screenshot 2024-10-31 at 1 05 27 PM

**Comment by dannymcclain** [comment]
@luacmartins I definitely think we should distinguish those somehow. Right now that list is basically unusable. There's a million ways we could do that, here are a few ideas:

image

I think I lean towards the simple dot separator pattern: `in:#admins • Danny's Donuts`. But let's see what @Expensify/design thinks or if they have any other/better ideas.

**Comment by luacmartins** [comment]
Being discussed [here](https://expensify.slack.com/archives/C07HPDRELLD/p1730446558577429?thread_ts=1726090155.295889&cid=C07HPDRELLD)

**Review by rayane-d — CHANGES_REQUESTED** [review]

File: src/components/Search/SearchRouter/getQueryWithSubstitutions.ts
Comment (Kicu) [reply src/components/Search/SearchRouter/getQueryWithSubstitutions.ts:4]: I used to think we might need more info in this map, but since the current solution is working then I will simplify

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (Kicu) [reply src/components/Search/SearchRouter/SearchRouter.tsx:78]: done

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (Kicu) [reply src/components/Search/SearchRouter/SearchRouter.tsx:250]: I am following the slack thread for this and will ask there

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (Kicu) [reply src/components/Search/SearchRouter/SearchRouter.tsx:289]: Good point. I used the card generating function to not add new code and it looks nice

**Comment by Kicu** [comment]
@luacmartins @rayane-djouah I addressed I believe all the comments, with 2 special notes:
 - about report/room names, there is a conversation on slack where it was decided to leave them as they are for now, because they will get a better display in future across the app, @JmillsExpensify confirmed this
 - about something looking differently in SearchPage - my next (and hopefully last) step is to move all this functionality to SearchPage results as well and share the code, so that in there everything looks and behaves the same in regards to autocomplete. That will come in a separate PR

Please re-check.

File: src/components/Search/SearchRouter/getQueryWithSubstitutions.ts
Comment (Kicu) [reply src/components/Search/SearchRouter/getQueryWithSubstitutions.ts]: I went over all the types and simplifed them a little because we had some redundancies. Now they have a better name

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (dannymcclain) [reply src/components/Search/SearchRouter/SearchRouter.tsx]: Personally I like the idea of doing the last 4 digits with the dot separator: `Card name • 1234`

I _think_ we're leaning towards a similar solution for rooms that have the same name across workspaces (i.e. `#social`, `#admins`, etc.

Definitely want to hear from the rest of @Expensify/design though.

Also, does the query syntax for this need to be `cardID:` or could it just be `card:`?

Comment (shawnborton) [reply src/components/Search/SearchRouter/SearchRouter.tsx]: Interesting, for this idea @dannymcclain:

> Personally I like the idea of doing the last 4 digits with the dot separator:

The way I understand it is that if you don't give the card a nickname, the card's default name includes the last four. But if you give it a custom nickname, you can overwrite the card name and put whatever you want there.

I think I lean towards just doing something that uses the card name exactly as it appears in the cards table, so if there is a nickname, use that. But if not, use the default name (e.g. `Expensify Card - 1234`

This way we don't have to worry about any custom naming logic here, we just use whatever is in the table. Thoughts?

Comment (dannymcclain) [reply src/components/Search/SearchRouter/SearchRouter.tsx]: Ahh yeah ok that makes sense! I didn't realize it worked that way. Totally down for that approach in that case!

Comment (Kicu) [reply src/components/Search/SearchRouter/SearchRouter.tsx]: So I tried to copy the logic that generates names for on wallet Page, and here's how this looks (please ignore that there is more cards - this is fake dummy data):

Screenshot 2024-11-04 at 16 37 17

Does this make more sense? As far as I understand correctly the logic on Wallet is as follows:
 - if we added a title to the card - use it
 - otherwise use property `card.bank` which I'm guessing would be the bank name.

There is no 4 last digits shown in this case, but I can add them.

Comment (Kicu) [reply src/components/Search/SearchRouter/SearchRouter.tsx]: and @dannymcclain it is possible to changed `cardID:` into `card:` but it requires a bit of work and would best be coordinated with backend.
@luacmartins if making this change it would be best to switch it all across the grammar I think

Comment (shawnborton) [reply src/components/Search/SearchRouter/SearchRouter.tsx]: > There is no 4 last digits shown in this case, but I can add them.

What about the Expensify Cards table? Can you show us what some of those entries look like, especially ones without nicknames?

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (luacmartins) [reply src/components/Search/SearchRouter/SearchRouter.tsx:252]: TLDR this will be implemented as a holistic approach throughout the app, so we decided to do nothing for now.

**Review by luacmartins — CHANGES_REQUESTED** [review]
Thanks for addressing the comments. I noticed two small issues when testing this PR:

1. We show the email address in the search bar when searching by display name

https://github.com/user-attachments/assets/004cfb9b-1ebf-4210-a248-48ae6f987145

2. Selecting a tax rate shows a duplicated value in the search bar

https://github.com/user-attachments/assets/6c38847b-5774-456e-b968-15b6271fd4fe

**Comment by Kicu** [comment]
@luacmartins 
Taxes are simple to fix - I will just run a `uniq` function over the final names array and it should work.

The email problem is deeper. Please keep in mind that the whole mechanism of `substitutionMap` is not yet implemented for this specific input on results page. This was something I was hoping to do in a separate PR seeing as this one is pretty big.

So right now we display email in the input, because if you hit ENTER again, (or change something else in query and then hit enter) this `email` gets send to backend and we get results back.
If at this point we'd display `names`, then that input would basically stop working, because it is unable to connect the name to email, and `from:asdf asdf` would get sent to backend, giving us back no results.

So can I suggest we wait with this one? My expectations is that I'd need 2-3 days after this is merged to bring full autocomplete into SearchPage results header.

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (Kicu) [reply src/components/Search/SearchRouter/SearchRouter.tsx]: @shawnborton sorry for my ignorance, but having never worked with the card related code I'm not sure I know how to properly simulate expensify cards 😅

Here are screenshots of 2 different methods of generating name
### 1. same as in wallet page
screen-cards1
### 2. using card `title` or `bank` 
screen-cards2

The 3rd one from the top is what I believe is an "expensify card" since I set the bank name to `Expensify Bank`.
If that doesn't answer the question, then I need @luacmartins help

Comment (shawnborton) [reply src/components/Search/SearchRouter/SearchRouter.tsx]: Oh man I am torn here... the Wallet naming seems to make the most sense to me at first, but then we run into the issue where we use the generic name for Expensify Cards and thus all unnamed Expensify Cards would look the same right? Let's see if @luacmartins has any thoughts here.

**Comment by luacmartins** [comment]
>So can I suggest we wait with this one? My expectations is that I'd need 2-3 days after this is merged to bring full autocomplete into SearchPage results header.

Yea, that's fine. Let's address it in a separate PR.

**Comment by Kicu** [comment]
great, the taxes unique issue is fixed in last commit, the whole thing can be reviewed again @rayane-djouah 
The last open issue is the naming of the cards: https://github.com/Expensify/App/pull/51633#discussion_r1828961475

But I think it shouldn't block this PR, and we can introduce any changes to card names in another PR if design team lands on a decision.

**Comment by luacmartins** [comment]
Agree! Let's continue testing @rayane-djouah !

**Review by luacmartins — DISMISSED** [review]
LGTM and tests well! Nice job. @rayane-djouah all yours

**Review by rayane-d — CHANGES_REQUESTED** [review]

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (Kicu) [reply src/components/Search/SearchRouter/SearchRouter.tsx:166]: yes, but I suggest I do this in a separate PR, similarly how i described it here: https://github.com/Expensify/App/pull/51633#issuecomment-2456503862

**Comment by Kicu** [comment]
@rayane-djouah 
> I noticed that the autocomplete options for "from," "to," and "in" are not sorted by most recent,

About this I'm not sure how they should be sorted. This sort order is already merged on main and I didn't really change the logic of how the options are generated.

**Review by rayane-d — COMMENTED** [review]
> I'm not sure how they should be sorted.

We could use a similar approach to what is implemented in the filter selector pages:

https://github.com/Expensify/App/blob/0a8f8976d62b80e4c07b453e89cd1f4573273a5c/src/components/Search/SearchFiltersParticipantsSelector.tsx#L46-L76

https://github.com/Expensify/App/blob/0a8f8976d62b80e4c07b453e89cd1f4573273a5c/src/components/Search/SearchFiltersChatsSelector.tsx#L64-L78

The sorting for `reports` and `personalDetails` is handled in the `getOptions` function here:

https://github.com/Expensify/App/blob/0a8f8976d62b80e4c07b453e89cd1f4573273a5c/src/libs/OptionsListUtils.ts#L1816-L1833

https://github.com/Expensify/App/blob/0a8f8976d62b80e4c07b453e89cd1f4573273a5c/src/libs/OptionsListUtils.ts#L1892-L1894

**Comment by luacmartins** [comment]
I think those are NAB and we can address them in a follow up. Up to you @Kicu

**Comment by luacmartins** [comment]
Sounds good. Let's push this PR forward then and address those in a follow up. @rayane-djouah

**Comment by luacmartins** [comment]
Reviewing it again now

**Review by luacmartins — APPROVED** [review]
LGTM and tests well

**Review by rayane-d — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/user-attachments/assets/2d60bf5e-9cea-4f55-8a79-26344303ac3c



</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/user-attachments/assets/48e656b2-25f4-4ac3-8862-c2f063701dff



</details>

<details>
<summary>iOS: Native</summary>


https://github.com/user-attachments/assets/acc658cf-b261-42c5-a4ec-7e7e4c0ddcee



</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/user-attachments/assets/01815209-2b49-4231-9c32-006a273f337d


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/user-attachments/assets/e3f624c4-30f0-451a-8609-72e3ac7f659b



</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/52c3cc67-da9a-4ae1-9680-f9aace51a3c4




</details>

**Review by luacmartins — APPROVED** [review]

**Comment by luacmartins** [comment]
Gonna merge this one. @Kicu let's address the comments above in a follow up!

**Comment by Kicu** [comment]
will link when I have a PR fixing all these things 👍

**Comment by Kicu** [comment]
@ntdiary I will keep this in mind when doing the next PR with some polish items 👍

**Pull request merged**
